### ISSUE TITLE
Add NonEmptyVec::push

### ISSUE THREAD

**Issue by jyn514** [lzutao/rust-oom#5]
And in general any methods that only allow adding to the vec but not removing. Otherwise it's really just a `Box<NonEmptySlice>` which doesn't seem to pull it's weight.

**Comment by tesuji** [comment]
Thanks for the interests.
I thought about adding `NonEmptyVec::push` or so but I don't particularly find the use case for it.

Also, `push` and `insert` are fallible operations, I am not sure how to cleanly handle them or just
mimic the std Vec API.

**Comment by jyn514** [comment]
When you say fallible you mean because the OS could run out of memory? It seems fine to panic in that case.

**Comment by tesuji** [comment]
And what about `insert` out of length ? I could return an `Result<(), T>` but it doesn't feel nice.

**Comment by tesuji** [comment]
~How about adding a `as_mut_vec -> &mut Vec`? After all  I don't know if the use case for push is big enough.~

it is a bad idea. People could use `pop` on `&mut Vec<T>` and the whole contract by `NonEmptyVec` is gone.

**Comment by tesuji** [comment]
So I thought a bit and found that I need more evident about the use case for `push`.
If the OP has more than 10 thumps up, and there is a concrete example that `push` is semantically useful,
I could change my mind and add that API.

**Comment by jyn514** [comment]
I don't feel very strongly about it, it just seemed an obvious addition to the API.

**Comment by tesuji** [comment]
Yeah, the current inner implementation allows use easily adding `push` method to `NonEmptyVec`.
But just because we can does not make it through. It has to be useful unless we want APIs
that nobody uses.

Alright, back to your concern in the first post:

> Otherwise it's really just a `Box<NonEmptySlice>` which doesn't seem to pull it's weight.

Maybe. But `NonEmptyVec` has inner capacity field, it should be really cheap to turn an instance of 
`NonEmptyVec<T>` into `Vec<T>` and back, unlike `Box<T>`.
